rclcpp/test/rclcpp/node_interfaces/test_node_parameters.cpp

```diff
@@ -197,5 +197,124 @@ TEST_F(TestNodeParameters, add_remove_parameters_callback) {
+        if (param.get_name() == "param1") {
+          parameters.emplace_back("param2", 2.0);
+        }
+      }
+    };
+
+    auto handle = node_parameters->add_pre_set_parameters_callback(callback);
+    auto result = node_parameters->set_parameters(parameters_to_be_set);
+    // we expect the result size to be same as the original "parameters_to_be_set"
+    // since the pre set parameter callback will set the modified param list atomically.
+    ASSERT_EQ(1u, result.size());
+    EXPECT_TRUE(result[0].successful);
+    EXPECT_TRUE(node_parameters->has_parameter("param1"));
+    EXPECT_EQ(node_parameters->get_parameter("param1").get_value<double>(), 1.0);
+    EXPECT_TRUE(node_parameters->has_parameter("param2"));
+    EXPECT_EQ(node_parameters->get_parameter("param2").get_value<double>(), 2.0);
+
+    EXPECT_NO_THROW(node_parameters->remove_pre_set_parameters_callback(handle.get()));
+
+    RCLCPP_EXPECT_THROW_EQ(
+        node_parameters->remove_pre_set_parameters_callback(handle.get()),
+        std::runtime_error("Pre set parameter callback doesn't exist"));
+  }
+
+  // the result will be unsuccessful if the pre set callback makes parameter list empty
+  {
+    auto callback = [](std::vector<rclcpp::Parameter> &parameters) {
+      parameters = {};
+    };
+
+    std::string reason = "parameter list cannot be empty, this might be due to "
+                  "pre_set_parameters_callback modifying the original parameters list";
+    auto handle = node_parameters->add_pre_set_parameters_callback(callback);
+    auto results = node_parameters->set_parameters(parameters_to_be_set);
+    EXPECT_FALSE(results[0].successful);
+    EXPECT_EQ(results[0].reason, reason);
+    EXPECT_NO_THROW(node_parameters->remove_pre_set_parameters_callback(handle.get()));
+
+    RCLCPP_EXPECT_THROW_EQ(
+        node_parameters->remove_pre_set_parameters_callback(handle.get()),
+        std::runtime_error("Pre set parameter callback doesn't exist"));
+  }
+}
+
+TEST_F(TestNodeParameters, add_remove_post_set_parameters_callback) {
+  rcl_interfaces::msg::ParameterDescriptor param1_descriptor;
+  param1_descriptor.name = "double_parameter1";
+  param1_descriptor.type = rcl_interfaces::msg::ParameterType::PARAMETER_DOUBLE;
+  param1_descriptor.read_only = false;
+
+  rcl_interfaces::msg::ParameterDescriptor param2_descriptor;
+  param2_descriptor.name = "double_parameter2";
+  param2_descriptor.type = rcl_interfaces::msg::ParameterType::PARAMETER_DOUBLE;
+  param2_descriptor.read_only = false;
+
+  node_parameters->declare_parameter(
+      "param1", rclcpp::ParameterValue(0.0),param1_descriptor, false);
+  node_parameters->declare_parameter(
+      "param2", rclcpp::ParameterValue(0.0),param2_descriptor, false);
+
+  double variable_tracking_param1_internally = node_parameters->get_parameter("param1").get_value<double>();
+  double variable_tracking_param2_internally = node_parameters->get_parameter("param2").get_value<double>();
```

Comment:
I think the following is equivalent:
```suggestion
   double variable_tracking_param1_internally = node_parameters->declare_parameter(
      "param1", rclcpp::ParameterValue(0.0),param1_descriptor, false);
  double variable_tracking_param2_internally = node_parameters->declare_parameter(
      "param2", rclcpp::ParameterValue(0.0),param2_descriptor, false);
```

Reply:
However, it would also be nice to test that declaring parameters also triggers the callback.